File: day4/part2.js
import * as fs from "fs";

// 2-4,6-8
// 2-3,4-5
// 5-7,7-9
// 2-8,3-7
// 6-6,4-6
// 2-6,4-8

export default function solve() {
  let count = 0;
  for (const line of fs.readFileSync("input.txt", "utf-8").split(/\r?\n/)) {
    const pairs = line.split(",");
    const p1 = pairs[0].split("-");
    const p2 = pairs[1].split("-");
    const a = +p1[0];
    const b = +p1[1];
    const c = +p2[0];
    const d = +p2[1];
    
    if ((a >= c && a <= d)
      || (b >= c && b <= d)
      || (c >= a && c <= b)
      || (d >= a && d <= b)) {
      count++;
    }
  }

  console.log(count);
}

File: day5/Part2.cs
using System.Text;

public class Part2
{
  public void Solve()
  {
    // Process Lines
    var stackLines = new List<string>();
    var opsLines = new List<string>();
    var cols = 0;
    foreach (var line in File.ReadLines("input.txt"))
    {
      if (line.Length == 0) continue;

      if (line[0] == 'm') opsLines.Add(line);
      else if (line[1] == '1') cols = line.Split(" ", StringSplitOptions.RemoveEmptyEntries).Length;
      else stackLines.Add(line);
    }

    // Create Stacks
    var stacks = new List<List<char>>(Enumerable.Range(0, cols).Select(x => new List<char>()));
    for (int i = stackLines.Count - 1; i >= 0; i--)
    {
      var line = stackLines[i];

      for (int j = 0; j < line.Length; j++)
      {
        if (line[j] == '[')
        {
          stacks[j / 4].Add(line[j + 1]);
        }
      }
    }

    // Performs ops
    foreach (var op in opsLines)
    {
      var nums = op.Split(" ").Where(x => char.IsDigit(x, 0)).Select(int.Parse).ToList();

      var delStack = stacks[nums[1] - 1];
      var addStack = stacks[nums[2] - 1];
      var ctr = nums[0];
      var idx = delStack.Count - ctr;

      var stack = delStack.Skip(idx);
      addStack.AddRange(stack);
      delStack.RemoveRange(idx, ctr);
    }

    var builder = new StringBuilder();
    foreach (var stack in stacks)
    {
      builder.Append(stack.Last());
    }

    System.Console.WriteLine(builder.ToString());
  }
}